### PR TITLE
(PUP-10041) Convert puppet module tool to http client

### PR DESCRIPTION
**Pull request by joshcooper**
* Updates forge related unit tests to use webmock
* Adds a forge integration test to verify SSL connections with the system CA store
* Changes puppet module tool to use http client instead of Puppet::Util::HttpProxy#get_http_object
* Only file metadata and content from http(s) file sources use Puppet::Util::HttpProxy to make Net::HTTP connections.

### CODE REVIEW

File: lib/puppet/forge/repository.rb [hunk 1]
```diff
@@ -15,38 +15,52 @@ class Repository
...
     def make_http_request(path, io = nil)
-      request = get_request_object(@uri.path.chomp('/')+path)
-      Puppet.debug "HTTP GET #{@host}#{request.path}"
-      return read_response(request, io)
+      raise ArgumentError, "Path must start with forward slash" unless path.start_with?('/')
+      begin
+        str = @uri.to_s
+        str.chomp!('/')
+        str += path
+        uri = URI(Puppet::Util.uri_encode(str))
```
Comment (joshcooper) [inline lib/puppet/forge/repository.rb]: Need to think about passing the `uri_encoded` string to `URI`.

File: lib/puppet/forge/repository.rb [hunk 2]
```diff
@@ -15,38 +15,52 @@ class Repository
...
+        http = Puppet.runtime['http']
+        response = http.get(uri, headers: headers, user: user, password: password, ssl_context: @ssl_context)
+        io.write(response.body) if io.respond_to?(:write)
+        response
+      rescue => e
+        case e.cause
+        when Puppet::SSL::CertVerifyError
+          raise SSLVerifyError.new(:uri => @uri.to_s, :original => e.cause)
+        when Puppet::SSL::SSLError
+          raise e.cause
```
Comment (joshcooper) [inline lib/puppet/forge/repository.rb]: This seems a little suspect. The http client API only raises `Puppet::HTTP::HTTPErrors` and derived exceptions. But here we need to differentiate between 1) cert verification error, 2) general ssl error, 3) non-ssl error. Peeking at the cause is cheating, as it's not part of the API. We could:

* Have the client raise `Puppet::SSL::SSLErrors` in addition to `Puppet::HTTP::HTTPError` and subclasses
* Create `Puppet::HTTP::SSLError` to mirror their `Puppet:SSL` counterparts

I'm thinking the former is better, but tests need to be added.

Comment (joshcooper) [reply lib/puppet/forge/repository.rb]: I implemented the former in https://github.com/puppetlabs/puppet/pull/7864/commits/aecceef1eef9afc9030585870560902d3776acff, but would be good to discuss.

File: lib/puppet/forge/repository.rb (hunk 1 above)
Comment (joshcooper) [reply lib/puppet/forge/repository.rb]: This is correct, we need to encode the string prior to passing to the URI parser. So this works:

```
irb(main):002:0> URI("http://foo/bar%20baz")
=> #<URI::HTTP http://foo/bar%20baz>
```

But this does not:

```
irb(main):003:0> URI("http://foo/bar baz")
URI::InvalidURIError: bad URI(is not URI?): http://foo/bar baz
```

**Comment by joshcooper** [comment]
jenkins please test this

**Comment by joshcooper** [comment]
@scotje @jpartlow We're changing how the PMT makes HTTP requests to the forge so that it behaves consistently with all of the puppet settings, etc. This PR preserves the encoding logic that was there before (at least that's the intent). We use `Puppet::Util.uri_encode` to URL encode the request path. So `+` and ` ` are always converted to `%20`. And query parameters are encoded the same way they were before.

I've tested this using `module_groups=base+pe` as that was an issue we had previously. Are there other combinations I should try? Are there some PE specific things I should try? I believe the behavior of using the forge authentication token and user/password encoded in the URL should still work.

**Comment by jpartlow** [comment]
The pe_acceptance_tests use puppet module in a few spots. The ace server tests, some pre and post upgrade tests, and a specific check of module dependencies that uses 'puppet module list'

In PE, it looks like we configure the master's puppet.conf with 'modules_groups = base+pe_only'.

And puppet module is referenced in the PE docs for some workflows that assume a simple code environment where the user can just puppet module a module into place on their master and continue with the workflow.

**Comment by joshcooper** [comment]
I installed PE 2019.3.x and the puppet-agent build from this PR, and verified I could install PE only modules and the pagination worked correctly:

```
$ puppet module search stdlib --http_debug
...
<- "GET /v3/modules?query=stdlib&module_groups=base%20pe_only&limit=20&offset=360 HTTP/1.1\r\nX-Puppet-Version: 6.12.0\r\nUser-Agent: PMT/1.1.1 (v3; Net::HTTP) Puppet/6.12.0 Ruby/2.5.7-p206 (x86_64-linux)\r\nAccept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3\r\nAccept: */*\r\nHost: forgeapi.puppet.com\r\n\r\n"
...
NAME                                          DESCRIPTION                                                               AUTHOR              KEYWORDS
puppetlabs-stdlib                             Standard library of resources for Puppet modules.                         @puppetlabs         library stdlib standard stages
geoffwilliams-stdlib DEPRECATED               Standard library of resources for Puppet modules.                         @geoffwilliams
puppetlabs-netdev_stdlib                      Type definitions for Networking Device (NetDev) Standard Library          @puppetlabs         network
hardening-hardening_stdlib                    Shared functions for hardening modules                                    @hardening          hardening stdlib
juniper-netdev_stdlib_junos                   Junos Provider code for Networking Device (netdev stdlib) Library         @juniper            networking juniper netdevops junos
puppetlabs-ciscopuppet                        Cisco Puppet providers and types for NX-OS devices                        @puppetlabs         cisco nxos nx-os network
brocade-netdev_stdlib_nos                     Provider and Type definition (based on netdev_stdlib) for Brocade VDX...  @brocade            network netdev brocade nos
aristanetworks-netdev_stdlib_eos              NetDev Providers for Arista EOS                                           @aristanetworks     eos netdev arista network
womply-stdlib_aws                             AWS management helper functions.                                          @womply
mellanox-netdev_stdlib_mlnxos                 Provider definition for implementing Networking Device (netdev) Libra...  @mellanox           netdev mlnxos l2 l3 vlan lag lacp
helio-fluentbit                               Installs and manages Fluentbit for Debian and Ubuntu                      @helio
mellanox-netdev_ospf_stdlib                   Type definitions for Networking Device (netdev) OSPF Library              @mellanox           netdev ospf l3 area router route
puppetlabs-java                               Installs the correct Java package on various platforms.                   @puppetlabs         java stdlib runtime jdk jre
netdevops-netdev_stdlib                       Type definitions for Networking Device (netdev) Library                   @netdevops          networking netdevops
huawei-netdev_stdlib_ce                       Huawei Puppet types and providers for CE devices                          @huawei
h3c-netdev_stdlib                             H3C Puppet types and providers for devices                                @h3c
puppet-extlib                                 extlib provides functions out of scope puppetlabs/stdlib                  @puppet             voxpupuli extlib stdlib
puppet-autofs                                 Module for installing and managing autofs                                 @puppet             autofs automount mount
simp-aide                                     manages AIDE                                                              @simp               simp aide
trlinkin-validate_multi                       Validate that the first argument given can pass validation of at leas...  @trlinkin           parser dsl validate stdlib
icann-opendnssec                              OpenDNSSEC configuration for GSI infrastructure ICANN                     @icann
puppetlabs-tomcat                             Installs, deploys, and configures Apache Tomcat web services.             @puppetlabs         tomcat
simp-sssd                                     Manages SSSD                                                              @simp               simp sssd
puppetlabs-activemq                           Installs and configures ActiveMQ.                                         @puppetlabs         java activemq amqp stomp stdlib
brainsware-resources_deep_merge               provides a function deep-merges every memeber of one hash1 with hash2     @brainsware         function stdlib deep-merge
viniciusxavier-ssh                            This module manage SSH client and server.                                 @viniciusxavier
rdark-mungelib                                A stdlib of sorts, a collection of functions for munging, mangling an...  @rdark
ptomulik-ptlib                                Utility functions for puppet module developers                            @ptomulik           stdlib custom delete values param
... hundreds more ....
```

**Comment by jpartlow** [comment]
:+1: I don't think the acceptance suite is doing anything more than that.

**Comment by joshcooper** [comment]
I talked this over in slack with some folks and got 👍 so I'm going to merge.

**Pull request merged**